apis/python/src/tiledb/vector_search/object_readers/bioimage_reader.py

```diff
@@ -0,0 +1,210 @@
+
+    def get_partitions(self, partition_size: int = -1) -> List[BioImagePartition]:
+        if partition_size == -1:
+            partition_size = 1
+
+        if self.crop_config is None:
+            self.create_crop_config()
+
+        partitions = []
+        partition_id = 0
+        for start in range(0, self.num_vectors(), partition_size):
+            end = start + partition_size
+            if end > self.num_vectors():
+                end = self.num_vectors()
+            partitions.append(BioImagePartition(str(partition_id), crop_config=self.crop_config[start:end], level=self.level, id_start=start, id_end=end))
+            partition_id += 1
+
+        return partitions
+
+    def read_objects(self, partition: BioImagePartition) -> Tuple[OrderedDict, OrderedDict]:
+        import tiledb
+        import numpy as np
+        from tiledb.bioimg.openslide import TileDBOpenSlide
+
+
+        if self.crop_config is None:
+            self.create_crop_config()
+        ids_per_image = {}
+        i = 0
+        for id in range(partition.id_start, partition.id_end):
+            image_uri = partition.crop_config[i][0]
+            if image_uri in ids_per_image:
+                ids_per_image[image_uri].append(id)
+            else:
+                ids_per_image[image_uri] = [id]
+            i += 1
+
+        size = len(partition.crop_config)
+        images = np.empty(size, dtype="O")
+        shapes = np.empty(size, dtype="O")
+        external_ids = np.zeros(size, dtype=np.uint64)
+        i = 0
+        image_uris = np.empty(size, dtype="O")
+        locations = np.empty(size, dtype="O")
+        with tiledb.scope_ctx(ctx_or_config=self.config):
+            for image_uri, ids in ids_per_image.items():
+                print(f"Loading image: {image_uri}")
+                slide = TileDBOpenSlide(image_uri)
+                level_dimensions = slide.level_dimensions[self.level]
+                image = slide.read_region((0, 0), self.level, level_dimensions)
+                for id in ids:
+                    dim_0_start = self.crop_config[id][1][0][0]
+                    dim_0_end = min(dim_0_start + self.object_crop_shape[0], level_dimensions[1])
+                    dim_1_start = self.crop_config[id][1][0][1]
+                    dim_1_end = min(dim_1_start + self.object_crop_shape[1], level_dimensions[0])
+                    cropped_image = image[dim_0_start:dim_0_end, dim_1_start:dim_1_end]
+                    images[i] = cropped_image.flatten()
+                    shapes[i] = np.array(cropped_image.shape, dtype=np.uint32)
+                    external_ids[i] = id
+                    image_uris[i] = image_uri
+                    locations[i] = np.array([dim_0_start, dim_0_end, dim_1_start, dim_1_end], dtype=np.uint32)
+                    i += 1
+        
+        return {"image": images, "shape": shapes, "external_id": external_ids}, {"image_uri": image_uris, "location": locations, "external_id": external_ids}
```

Comment:
(Unsure) The naming here seems a little funny - should we make these plural if they are lists of objects?
```suggestion
        return {"images": images, "shapes": shapes, "external_ids": external_ids}, {"image_uris": image_uris, "locations": locations, "external_ids": external_ids}
```
Same question on the other objects returned in `read_objects_by_external_ids`.

Reply:
This is more about batching multiple objects in on call and it follows the TileDB read API convention.

For example, a TileDB array can have attributes (i.e. "image", "shape"). When reading slices of data the result is a dictionary with the attribute names as keys and values a numpy array of all matching result values. Hence the mapping "image": images(np.array of multiple images)